driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -150,13 +170,73 @@ private void buildTableMetadata(KeyspaceMetadata ksm, List<Row> cfRows, Map<Stri
                         cols = Collections.<String, ColumnMetadata.Raw>emptyMap();
                     }
                 }
-                TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+
+                TableMetadata tm = TableMetadata.build(ksm, cfRow, cols, cassandraVersion);
+                TableMetadata tmPrevious = ksm.tables.get(tm.getName());
+                if (tmPrevious == null) {
+                    ksm.add(tm);
+                    triggerOnTableAdded(tm);
```

Comment:
Sorry I wasn't clear at all in my last comment: currently _all_ schema change events are processed in the main executor _except_ in tryConnect(). What I meant is that I was in favor of leaving it that way, _even knowing that it's potentially a bad idea to let user code go in that executor_, because like Olivier said it will fatally block some of the driver's threads: e.g. the handle() method is executed by an IO thread, and blocking that one could be even worse.